pkg/router/template/configmanager/haproxy/client.go

```diff
@@ -0,0 +1,204 @@
+		maps:          make(map[string]*HAProxyMap),
+	}
+}
+
+// RunCommand executes a haproxy dynamic config API command and if present
+// converts the response as desired.
+func (c *Client) RunCommand(cmd string, converter Converter) ([]byte, error) {
+	glog.V(4).Infof("Running haproxy command: %q ...", cmd)
+	buffer, err := c.runCommandWithRetries(cmd, maxRetries)
+	if err != nil {
+		glog.Warningf("haproxy dynamic config API command %s failed with error: %v", cmd, err)
+		return nil, err
+	}
+
+	response := buffer.Bytes()
+	glog.V(4).Infof("haproxy command response: %q", string(response))
+	if converter == nil {
+		return response, nil
+	}
+
+	return converter.Convert(response)
+}
+
+// Execute runs a haproxy dynamic config API command.
+func (c *Client) Execute(cmd string) ([]byte, error) {
+	return c.RunCommand(cmd, nil)
+}
+
+// Reset resets any changes and clears the backends and maps.
+func (c *Client) Reset() {
+	c.backends = make([]*Backend, 0)
+	c.maps = make(map[string]*HAProxyMap)
+}
+
+// Commit flushes out any pending changes on all the backends and maps.
+func (c *Client) Commit() error {
+	for _, b := range c.backends {
+		if err := b.Commit(); err != nil {
+			return err
+		}
+	}
+
+	for _, m := range c.maps {
+		if err := m.Commit(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Backends returns the list of configured haproxy backends.
+func (c *Client) Backends() ([]*Backend, error) {
+	if len(c.backends) == 0 {
+		if backends, err := GetHAProxyBackends(c); err != nil {
+			return nil, err
+		} else {
+			c.backends = backends
+		}
+	}
+
+	return c.backends, nil
+}
+
+// FindBackend returns a specific haproxy backend if it is configured.
+func (c *Client) FindBackend(id string) (*Backend, error) {
+	if _, err := c.Backends(); err != nil {
+		return nil, err
+	}
+
+	for _, b := range c.backends {
+		if b.Name() == id {
+			return b, nil
+		}
+	}
+
+	return nil, fmt.Errorf("no backend found for id: %s", id)
+}
+
+// Maps returns the list of configured haproxy maps.
+func (c *Client) Maps() ([]*HAProxyMap, error) {
+	if len(c.maps) == 0 {
+		hapMaps, err := GetHAProxyMaps(c)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, v := range hapMaps {
+			c.maps[v.Name()] = v
+		}
+
+		return hapMaps, nil
+	}
+
+	mapList := make([]*HAProxyMap, len(c.maps))
+	i := 0
+	for _, v := range c.maps {
+		mapList[i] = v
+		i++
+	}
+
+	return mapList, nil
+}
+
+// FindMap returns a populated haproxy map.
+func (c *Client) FindMap(name string) (*HAProxyMap, error) {
+	if _, err := c.Maps(); err != nil {
+		return nil, err
+	}
+
+	if m, ok := c.maps[name]; ok {
+		return m, m.Refresh()
+	}
+
+	return nil, fmt.Errorf("no map found for name: %s", name)
+}
+
+// runCommandWithRetries retries a haproxy command upto the retry limit
+// if the error for the command is a retryable error.
+func (c *Client) runCommandWithRetries(cmd string, limit int) (*bytes.Buffer, error) {
+	retryAttempt := 0
+	for {
```

Comment:
use util.ExponentialBackoff or util.PollImmediate instead of inventing your own code for it.